src/main/java/uk/gov/hmcts/ccd/domain/service/callbacks/CallbackService.java

```diff
@@ -171,8 +180,55 @@ public void validateCallbackErrorsAndWarnings(final CallbackResponse callbackRes
         }
     }
 
+    protected void addPassThroughHeaders(final HttpHeaders httpHeaders) {
+        if (null != request && null != applicationParams
+            && null != applicationParams.getCallbackPassthruHeaderContexts()) {
+            applicationParams.getCallbackPassthruHeaderContexts().stream()
+                .forEach(context -> addPassThruContextValuesToHttpHeaders(httpHeaders, context));
+        }
+    }
+
+    private void storePassThroughHeadersAsRequestAttributes(final HttpHeaders httpHeaders,
+                                                              HttpServletRequest request) {
+        if (null != request && null != applicationParams
+            && null != applicationParams.getCallbackPassthruHeaderContexts()) {
+            applicationParams.getCallbackPassthruHeaderContexts().stream()
+                .filter(context -> StringUtils.hasLength(context) && null != httpHeaders.get(context))
+                .forEach(context -> {
+                    LOG.debug("Setting request ATTRIBUTE context <{}> to value: <{}>",
+                        context, httpHeaders.get(context));
+                    request.setAttribute(context, httpHeaders.get(context));
+                });
+        }
+    }
+
+    private void addPassThruContextValuesToHttpHeaders(HttpHeaders httpHeaders, String context) {
+        if (null != request.getAttribute(context)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Use request ATTRIBUTE context <{}>: value <{}>", context, request.getAttribute(context));
+            }
+            if (httpHeaders.containsKey(context)) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Removing headers context <{}>: value <{}>", context, httpHeaders.get(context));
+                }
+                httpHeaders.remove(context);
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Add headers context <{}>: value <{}>", context, request.getAttribute(context));
+            }
+            httpHeaders.add(context, request.getAttribute(context).toString());
+            // tidy up? remove the attribute
+            request.removeAttribute(context);
```

Comment:
If this implementation is complete, can delete the comment.

Reply:
Done